### PR TITLE
Migrate mdatagen tool from using yaml to koanf/mapstructure

### PR DESCRIPTION
**Pull request by dmitryax**
For consistency with other parts of the project. It also helps to further simplify the yaml unmarshalling and validation work going forward.

### CODE REVIEW

File: cmd/mdatagen/loader.go [hunk 1]
```diff
@@ -45,15 +50,13 @@ type metric struct {
 
 	// ExtendedDocumentation of the metric. If specified, this will
 	// be appended to the description used in generated documentation.
-	ExtendedDocumentation string `yaml:"extended_documentation"`
+	ExtendedDocumentation string `mapstructure:"extended_documentation"`
 
 	// Unit of the metric.
-	Unit string `yaml:"unit"`
+	Unit string
```
Comment (bogdandrutu) [inline cmd/mdatagen/loader.go:56]: Why removing the `tag`?

File: cmd/mdatagen/loader.go [hunk 2]
```diff
@@ -45,15 +50,13 @@ type metric struct {
...
 
-	// Raw data that is used to set Data interface below.
-	YmlData *ymlMetricData `yaml:"data" validate:"required"`
 	// Date is set to generic metric data interface after validating.
-	Data MetricData `yaml:"-"`
+	Data MetricData `validate:"required"`
```
Comment (bogdandrutu) [inline cmd/mdatagen/loader.go:59]: I think we should do a bit different here maybe and adopt something similar with receivers protocols, see https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/jaegerreceiver/config.go#L44

Idea user will write:

```yaml
metrics:
  system.cpu.time:
    description: Total CPU seconds broken down by different states.
    unit: s
    sum:
      monotonic: true
      aggregation: cumulative
    attributes: [freeFormAttribute, freeFormAttributeWithValue, enumAttribute]
``

Comment (dmitryax) [reply cmd/mdatagen/loader.go:59]: I was also thinking about this interface. The only concern is that it allows having several datatypes since it's set as a key now, so we need to put some more validation in place

File: cmd/mdatagen/loader.go (hunk 1 above)
Comment (dmitryax) [reply cmd/mdatagen/loader.go:56]: This must be parsed as `"unit"` key by default. Looks like we always specify it in other places. I'll bring it back

File: cmd/mdatagen/loader.go (hunk 2 above)
Comment (bogdandrutu) [reply cmd/mdatagen/loader.go:59]: We will have a "validation", much simpler than current logic I think, also much simpler for users to read what properties are accepted for specific types.

Comment (dmitryax) [reply cmd/mdatagen/loader.go:59]: Ok. I applied this approach in another PR https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/6463. I agree, it looks better now. I'll update this one after

**Comment by bogdandrutu** [comment]
Please rebase.

**Comment by dmitryax** [comment]
Depends on https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/6469

**Review by bogdandrutu — COMMENTED** [review]
Can you split the changes to the test files "extracting to files" in a separate PR?

**Comment by dmitryax** [comment]
@bogdandrutu sure, done https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/6470

File: cmd/mdatagen/main_test.go [hunk 3]
```diff
@@ -29,7 +29,7 @@ name: metricreceiver
 metrics:
   system.cpu.time:
     description: Total CPU seconds broken down by different states.
-    extended_description: Additional information on CPU Time can be found [here](https://en.wikipedia.org/wiki/CPU_time).
+    extended_documentation: Additional information on CPU Time can be found [here](https://en.wikipedia.org/wiki/CPU_time).
```
Comment (dmitryax) [inline cmd/mdatagen/main_test.go:32]: This change is required to pass new mapstructure validation. Submitted another PR https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/6473

File: cmd/mdatagen/loader.go [hunk 4]
```diff
@@ -100,20 +100,22 @@ type templateContext struct {
...
 }
 
-func loadMetadata(ymlData []byte) (metadata, error) {
-	var out metadata
-
-	// Unmarshal metadata.
-	if err := yaml.Unmarshal(ymlData, &out); err != nil {
-		return metadata{}, fmt.Errorf("unable to unmarshal yaml: %v", err)
+func loadMetadata(filePath string) (metadata, error) {
+	mdMap, err := config.NewMapFromFile(filePath)
```
Comment (bogdandrutu) [inline cmd/mdatagen/loader.go:104]: Use the configmapprovider please. See https://github.com/open-telemetry/opentelemetry-collector/pull/4502/

Comment (dmitryax) [reply cmd/mdatagen/loader.go]: done

**Comment by dmitryax** [comment]
Needs this https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/6502 to pass the build-and-test

**Comment by jpkrohling** [comment]
#6502 merged.

**Review by bogdandrutu — APPROVED** [review]
